### PR TITLE
Cavities report

### PR DESCRIPTION
**Pull request by rlavendomme**
Adapt unit cell code for the flood fill algorithm,
Add cavity sorting by volume function.
Add export list of cavities surface maps.
Add list of cavity in report.
Minor fixes.

### CODE REVIEW

**Review by jmaglic — CHANGES_REQUESTED** [review]
I left some comments for you but mostly this seems fine. One thing I've noticed is that with my unit test for testing the two-probe mode, all volume values except probe 1 core volume have slightly changed. I couldn't find the reason for this. Do you have any idea what happened? I feel that at least the vdW volume should have stayed the same. Did you make changes to how the size of the space is determined in two probe mode? Now that I think of it I think I saw something along those lines.

Before we merge I want to make an attempt at rewriting the vector sorting function (sortCavities or something like that) in a shorter and hopefully more elegant way. If I succeed I'll add a commit to this branch and let you look over it.

File: src/controller.cpp [hunk 1]
```diff
@@ -42,7 +42,6 @@ bool Ctrl::loadRadiusFile(){
   if(!current_calculation->readRadiusFileSetMaps(radius_filepath)){
     notifyUser("\nInvalid radii definition file!");
     notifyUser("\nPlease select a valid file or set radii manually.");
-    // shouldn't this return false? -JM
```
[the remark this answers is not in the document]
Comment (rlavendomme) [reply src/controller.cpp:45]: I thought the message for the user was self-explanatory `Please select a valid file or set radii manually` but I can add a comment about it.

File: src/base_event.cpp [hunk 2]
```diff
@@ -71,13 +71,13 @@ void MainFrame::OnCalc(wxCommandEvent& event){
 
   // write total surface map file if option is toggled
   if(getMakeSurfaceMap()){
-    Ctrl::getInstance()->exportSurfaceMap();
+    Ctrl::getInstance()->exportSurfaceMap(0);
```
[the remark this answers is not in the document]
Comment (rlavendomme) [reply src/base_event.cpp:74]: No particular reason. I use them interchangeably but we can stick to `true `and `false `for homogeneity. I'll change it.

File: src/model.cpp [hunk 3]
```diff
@@ -180,7 +188,12 @@ CalcReportBundle Model::calcVolume(){
   _data.success = true;
 
   auto start = std::chrono::steady_clock::now();
-  _cell.assignTypeInGrid(atomtree, getProbeRad1(), getProbeRad2(), optionProbeMode()); // assign each voxel in grid a type
+  bool error_cav = false;
+  _cell.assignTypeInGrid(atomtree, getProbeRad1(), getProbeRad2(), optionProbeMode(), error_cav); // assign each voxel in grid a type
+  if(error_cav){
+    Ctrl::getInstance()->notifyUser("\n\nWARNING: Maximum number of cavities reached. Not all cavities are reported.");
+    Ctrl::getInstance()->notifyUser("\nTo solve this, change probe radius (e.g. smaller probe 2 and/or larger probe 1).\n");
+  }
```
[the remark this answers is not in the document]
Comment (rlavendomme) [reply src/model.cpp:196]: I'll add a TODO comment so we can easily find all these small desired modifications later.

File: src/model.cpp [hunk 4]
```diff
@@ -191,17 +204,85 @@ CalcReportBundle Model::calcVolume(){
...
   end = std::chrono::steady_clock::now();
   _data.addTime(std::chrono::duration<double>(end-start).count());
 
+  sortCavities();
+
   return _data;
 }
 
+// sort cavities related vectors in _data in descending occupied volume order
+void Model::sortCavities(){
```
[the remark this answers is not in the document]
Comment (rlavendomme) [reply src/model.cpp:221]: I thought you would want to re-code that function. Feel free to make a better version.

**Comment by rlavendomme** [comment]
> 
> 
> One thing I've noticed is that with my unit test for testing the two-probe mode, all volume values except probe 1 core volume have slightly changed. I couldn't find the reason for this. Do you have any idea what happened? I feel that at least the vdW volume should have stayed the same. Did you make changes to how the size of the space is determined in two probe mode? Now that I think of it I think I saw something along those lines.
> 

The change you mentioned is probably due to the calibration of the voxel grid with the cartesian coordinate that was introduced in commit https://github.com/jmaglic/MoloVol/commit/03353efd9101e9538d058ef1a56c3b8a9774fd01
This is not directly related to this branch. At least for me, I have same volume values between the master branch and this development branch.

This calibration was done to simplify unit cell volume (and later surface) calculations. The values should have remained in a very small error range. 
Can you confirm that the difference is negligible or did you observe completely wrong values?

**Comment by jmaglic** [comment]
Just compiled the master branch and I can confirm that it's the same there, so I must have missed this change last time round. Seem to me quite minor deviations.

**Pull request merged**